src/main/kotlin/org/jitsi/jibri/capture/ffmpeg/executor/impl/AbstractFfmpegExecutor.kt

```diff
@@ -0,0 +1,150 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.jitsi.jibri.capture.ffmpeg.executor.impl
+
+import org.jitsi.jibri.capture.ffmpeg.executor.FfmpegExecutor
+import org.jitsi.jibri.capture.ffmpeg.executor.FfmpegExecutorParams
+import org.jitsi.jibri.capture.ffmpeg.util.ENCODING_KEY
+import org.jitsi.jibri.capture.ffmpeg.util.FfmpegFileHandler
+import org.jitsi.jibri.capture.ffmpeg.util.OutputParser
+import org.jitsi.jibri.capture.ffmpeg.util.WARNING_KEY
+import org.jitsi.jibri.sink.Sink
+import org.jitsi.jibri.util.NameableThreadFactory
+import org.jitsi.jibri.util.Tail
+import org.jitsi.jibri.util.Tee
+import org.jitsi.jibri.util.extensions.debug
+import org.jitsi.jibri.util.extensions.error
+import org.jitsi.jibri.util.pid
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.logging.Logger
+
+const val FFMPEG_RESTART_ATTEMPTS = 1
+/**
+ * [AbstractFfmpegExecutor] contains logic common to launching Ffmpeg across platforms.
+ * It is abstract, and requires a subclass to implement the
+ * [getFfmpegCommand] method to return the proper command.
+ */
+abstract class AbstractFfmpegExecutor(private val processBuilder: ProcessBuilder = ProcessBuilder()) : FfmpegExecutor {
+    private val logger = Logger.getLogger(AbstractFfmpegExecutor::class.qualifiedName)
+    private val ffmpegOutputLogger = Logger.getLogger("ffmpeg")
+    private val executor = Executors.newSingleThreadExecutor(NameableThreadFactory("AbstractFfmpegExecutor"))
+    /**
+     * The currently active (if any) Ffmpeg process
+     */
+    private var currentFfmpegProc: Process? = null
+    /**
+     * In order to both analyze Ffmpeg's output live and log it to a file, we'll
+     * tee its stdout stream so that both consumers can get all of its output.
+     */
+    private var ffmpegOutputTee: Tee? = null
+    /**
+     * We'll use this to monitor the stdout output of the Ffmpeg process
+     */
+    private var ffmpegTail: Tail? = null
+
+    init {
+        ffmpegOutputLogger.useParentHandlers = false
+        ffmpegOutputLogger.addHandler(FfmpegFileHandler())
+    }
+
+    /**
+     * Get the shell command to use to launch ffmpeg
+     */
+    protected abstract fun getFfmpegCommand(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink): String
+
+    override fun launchFfmpeg(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink): Boolean {
+        processBuilder.command(getFfmpegCommand(ffmpegExecutorParams, sink).split(" "))
```

Comment:
Probably not worth worrying about at this point, but I think this will misbehave if the path contains spaces.

Reply:
i actually have been thinking about this situation, because in the sipgw case the arguments are formatted differently and this method doesn't work well, so going to try and see if i can come up with a better way

Reply:
i cleaned this up in the sip gateway branch